### PR TITLE
Travis CI: optimize setting up containers

### PR DESCRIPTION
**Pull request by simison**
This is a follow up to parallelization improvements done in https://github.com/Automattic/jetpack/pull/9772 (p9dueE-lr-p2).

This splits up Node.js and PHP environments furthermore so that we don't setup all different WP versions, PHP setup and Node.js setup in every container.

Instead, we setup only what we need in each container.

- Add `language: node_js` to the two containers requiring Node.js and drop installing Node.js in PHP containers. Saves ~10sec from PHP containers and probably something also in Node.js containers since they now rely on Travis' ready-made containers.
- Let Travis pick Node.js version from [`.nvmrc` file](https://github.com/creationix/nvm#nvmrc) — like this we're always in sync and define the version in fewer places.
- Update Node.js version configured in `.nvmrc` from 8 LTS to 10 LTS — Version 10 should come with some speed improvements as well and in my testing it works great for regular Jetpack development.
- Remove unneeded `gem sass` and `gem compass` installs — dropping these will speed things up a bit.
- Configure PHP+PHPUnit setup in new `setup-travis.sh` for clarity instead of doing it for every container in `travis.yml` and drop `prepare-wordpress.sh` as redundant. We could run the setup script from `run-travis.sh`, but when ran from Travis config using `before_script`, we'll get a nice timer in Travis logs:
    image
- Don't clone 3 different types of WordPress in each container; clone only the branch we're going to test in that container.
- Remove `yarn install` since Travis does it for us in Node.js containers when it detects `yarn.lock` ([docs](https://docs.travis-ci.com/user/languages/javascript-with-nodejs#travis-ci-supports-yarn))

Average savings should be something like 10-30sec depending on a container.

As a further improvement in another PR, `yarn lint` should be running together with tests; since linter takes only 12s, it doesn't make sense to have its own container for it.

### Testing

- Does CI 💚 ? Is it faster?
- Do typical `yarn build` etc commands work with Node.js 10?

### CODE REVIEW

File: tests/setup-travis.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	composer global require "phpunit/phpunit=4.8.*" --no-suggest
+fi
+
+mysql -e "set global wait_timeout = 3600;"
+
+# Prepare a developer checkout of WordPress
+mysql -u root -e "CREATE DATABASE wordpress_tests;"
+
+echo "Preparing WordPress from \"$WP_BRANCH\" branch...";
+case $WP_BRANCH in
```
Comment (simison) [inline tests/setup-travis.sh]: This case was previously wrapped in a loop to install all 3 different branches:
```
for WP_SLUG in 'master' 'latest' 'previous'; do
...
```

File: .travis.yml [hunk 2]
```diff
@@ -63,28 +65,12 @@ branches:
...
-    - nvm install 8
-    - nvm use 8
-    - ./tests/prepare-wordpress.sh
-    - mysql -e "set global wait_timeout = 3600;"
+  - export PLUGIN_SLUG=$(basename $(pwd))
+  - export PATH="$HOME/.composer/vendor/bin:$PATH"
+  - ./tests/setup-travis.sh
 
-script:
-    - echo "Running with the following versions:"
```
Comment (simison) [inline .travis.yml]: `phpunit --version` is now part of `run-travis.sh`

Travis does `php -v` for us anyway when it spins up a PHP container:
image

File: .travis.yml [hunk 3]
```diff
@@ -63,28 +65,12 @@ branches:
...
-            composer global require "phpunit/phpunit=5.7.*"
-        elif [[ ${TRAVIS_PHP_VERSION:0:3} != "5.2" ]]; then
-            composer global require "phpunit/phpunit=4.8.*"
-        fi
-    - nvm install 8
-    - nvm use 8
-    - ./tests/prepare-wordpress.sh
-    - mysql -e "set global wait_timeout = 3600;"
+  - export PLUGIN_SLUG=$(basename $(pwd))
+  - export PATH="$HOME/.composer/vendor/bin:$PATH"
```
Comment (simison) [inline .travis.yml]: We want to inherit these constants to `run-travis.sh`, thus they should be here and not in `setup-travis.sh`

File: .travis.yml [hunk 4]
```diff
@@ -41,8 +41,10 @@ matrix:
   - php: "5.2"
     env: WP_MODE=multi  WP_BRANCH=master
     dist: precise
-  - env: WP_TRAVISCI="yarn lint"
-  - env: WP_TRAVISCI="yarn test-dangerci-and-adminpage"
+  - language: node_js
+    env: WP_TRAVISCI="yarn lint"
+  - language: node_js
```
Comment (simison) [inline .travis.yml]: Using `language:node_js` will let Travis pick us a Node.js container based on the version we have in `.nvmrc`

File: .travis.yml [hunk 5]
```diff
@@ -41,8 +41,8 @@ matrix:
   - php: "5.2"
     env: WP_MODE=multi  WP_BRANCH=master
     dist: precise
-  - env: WP_TRAVISCI="yarn lint"
-  - env: WP_TRAVISCI="yarn test-dangerci-and-adminpage"
+  - language: node_js
```
Comment (simison) [inline .travis.yml]: Using `language:node_js` will let Travis pick us a Node.js container based on the version we have in `.nvmrc`.

File: package.json [hunk 6]
```diff
@@ -45,8 +45,7 @@
     "add-textdomain": "wpi18n addtextdomain --textdomain=jetpack --glob-pattern='!(docker|node_modules|tests|tools|vendor){*.php,**/*.php}'",
     "test-gui": "NODE_ENV=test NODE_PATH=tests:_inc/client:node_modules/@automattic/dops-components/client tests/runner.js gui",
     "precommit": "node bin/pre-commit-hook.js",
-    "test-modules": "NODE_ENV=test NODE_PATH=tests:_inc/client:node_modules/@automattic/dops-components/client tests/runner.js modules",
-    "test-dangerci-and-adminpage": "yarn danger ci && yarn test-adminpage"
```
Comment (simison) [inline package.json]: I didn't see this `test-dangerci-and-adminpage` being used anywhere else except in Travis scripts so for the sake of simplicity, I've removed it and I'm running those two tasks directly instead.

**Comment by kraftbj** [comment]
Comparing the Travis run time for this compared to other PRs, there is a reduction of about two minutes.

I'd like someone who is more current with groking our Travis setup to give it the formal review, but the changes make sense to me.

**Review by brbrr — DISMISSED** [review]
Changes looks great! Thanks for speeding-up Jetpack.

One thing which isn't clear to me is how updating node v8->v10 will affect our JS code. I'd assume it should be fine, might adding a note in the changelog and/or test instructions will help bring enough attention to this.

File: .nvmrc [hunk 7]
```diff
@@ -1 +1 @@
-lts/carbon
+10
```
[the remark this answers is not in the document]
Comment (simison) [reply .nvmrc]: It does improve speed little bit and would bring Jetpack in-par with Calypso development. That's a big deal for folks who use `nvm` version auto-switching based on this file.

I've been doing Jetpack development with Node.js 10 for a long time already without issues but would be happy if folks who do the final build would confirm that their build scripts function with the change. (ping @jeherve)

**Comment by simison** [comment]
Rebased, good for a final review again. :thumbsup:

**Review by brbrr — APPROVED** [review]
🐑 🇮🇹 👍

**Review by kraftbj — APPROVED** [review]
This looks to be really helpful on advancing our Travis setup. The Node change is notable, but it makes sense to me to attempt to match Calypso. If we run into problems down the way, Jetpack should adjust to Node 10 vs keeping it pinned to `lts/carbon` IMO.

**Pull request merged**